lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -138,20 +135,19 @@ static ArrayAttr filterAnnotations(MLIRContext *ctxt, ArrayAttr annotations,
     return ArrayAttr::get(ctxt, retval);
   for (auto opAttr : annotations) {
     if (auto subAnno = opAttr.dyn_cast<SubAnnotationAttr>()) {
-      // Check for overlap of the two ranges
-      if ((field.minFieldID >= subAnno.getMinFieldID() &&
-           field.minFieldID <= subAnno.getMaxFieldID()) ||
-          (field.maxFieldID >= subAnno.getMinFieldID() &&
-           field.minFieldID <= subAnno.getMinFieldID())) {
-        auto newMin =
-            srcType.rootChildFieldID(subAnno.getMinFieldID(), field.index);
-        auto newMax =
-            srcType.rootChildFieldID(subAnno.getMinFieldID(), field.index);
-        if (newMin.first == 0 && newMax.first == 0)
+      // Check whether the annotation falls into the range of the current field.
+      if (subAnno.getFieldID() >= field.fieldID &&
+          subAnno.getFieldID() <= field.fieldID + field.type.getMaxFieldID()) {
+        if (auto newFieldID = subAnno.getFieldID() - field.fieldID) {
+          // If the target is a subfield/subindex of the current field, create a
+          // new sub-annotation with a new field ID.
+          retval.push_back(SubAnnotationAttr::get(ctxt, newFieldID,
+                                                  subAnno.getAnnotations()));
+        } else {
+          // Otherwise, if the current field is exactly the target, degenerate
+          // the sub-annotation to a normal annotation.
           retval.push_back(subAnno.getAnnotations());
-        else
-          retval.push_back(SubAnnotationAttr::get(
-              ctxt, newMin.first, newMax.first, subAnno.getAnnotations()));
+        }
```

Comment:
This looks good and is a huge simplification.  One thing: currently this will drop `subAnno<fieldID = 0, ...>`.  I _think_ this should result in the annotation getting applied to everything. What do you think?

Reply:
`subAnno<fieldID = 0, {annotations}>` is equivalent to `{annotations}` - applying the annotations to all components. I think this should be the case?

Reply:
Can you verify that it does do this and add a test? I grabbed this locally and it was getting dropped.

See:
```mlir
module  {
  firrtl.circuit "Foo"   {
    firrtl.module @Foo() {
      %a = firrtl.wire  {
        annotations = [#firrtl.subAnno<fieldID = 0, {a}>]
      } : !firrtl.vector<vector<uint<1>, 2>, 2>
    }
  }
}
```

With `circt-opt -pass-pipeline='firrtl.circuit(firrtl-lower-types)' Foo.mlir`:

```mlir
module  {
  firrtl.circuit "Foo"   {
    firrtl.module @Foo() {
      %a_0_0 = firrtl.wire  : !firrtl.uint<1>
      %a_0_1 = firrtl.wire  : !firrtl.uint<1>
      %a_1_0 = firrtl.wire  : !firrtl.uint<1>
      %a_1_1 = firrtl.wire  : !firrtl.uint<1>
    }
  }
}
```

Reply:
Wow, thanks for catching this...

Reply:
Fixed this and added a test case. The MemOp handler has a similar issue, which is also fixed. Please take a look whether I missed anything, thanks!